core/src/trezor/wire/thp/channel.py

```diff
@@ -0,0 +1,319 @@
+        _, _, payload_length = ustruct.unpack(">BHH", packet)
+        self.expected_payload_length = payload_length
+        packet_payload = memoryview(packet)[INIT_HEADER_LENGTH:]
+
+        # If the channel does not "own" the buffer lock, decrypt first packet
+        # TODO do it only when needed!
+        # TODO FIX: If "_decrypt_single_packet_payload" is implemented, it will (possibly) break "decrypt_buffer" and nonces incrementation.
+        # On the other hand, without the single packet decryption, the "advanced" buffer selection cannot be implemented
+        # in "memory_manager.select_buffer", because the session id is unknown (encrypted).
+
+        # if control_byte.is_encrypted_transport(ctrl_byte):
+        #   packet_payload = self._decrypt_single_packet_payload(packet_payload)
+
+        self.buffer = memory_manager.select_buffer(
+            self.get_channel_state(),
+            self.buffer,
+            packet_payload,
+            payload_length,
+        )
+        await self._buffer_packet_data(self.buffer, packet, 0)
+
+        if __debug__:
+            log.debug(__name__, "handle_init_packet - payload len: %d", payload_length)
+            log.debug(__name__, "handle_init_packet - buffer len: %d", len(self.buffer))
+
+    async def _handle_cont_packet(self, packet: utils.BufferType) -> None:
+        if __debug__:
+            log.debug(__name__, "handle_cont_packet")
+        if not self.is_cont_packet_expected:
+            raise ThpError("Continuation packet is not expected, ignoring")
+        await self._buffer_packet_data(self.buffer, packet, CONT_HEADER_LENGTH)
+
+    def _decrypt_single_packet_payload(
+        self, payload: utils.BufferType
+    ) -> utils.BufferType:
+        # crypto.decrypt(b"\x00", b"\x00", payload_buffer, INIT_DATA_OFFSET, len(payload))
+        return payload
+
+    def decrypt_buffer(
+        self, message_length: int, offset: int = INIT_HEADER_LENGTH
+    ) -> None:
+        noise_buffer = memoryview(self.buffer)[
+            offset : message_length - CHECKSUM_LENGTH - TAG_LENGTH
+        ]
+        tag = self.buffer[
+            message_length
+            - CHECKSUM_LENGTH
+            - TAG_LENGTH : message_length
+            - CHECKSUM_LENGTH
+        ]
+        if utils.DISABLE_ENCRYPTION:
+            is_tag_valid = tag == crypto.DUMMY_TAG
+        else:
+            key_receive = self.channel_cache.get(CHANNEL_KEY_RECEIVE)
+            nonce_receive = self.channel_cache.get_int(CHANNEL_NONCE_RECEIVE)
+            auth_data = self.get_handshake_hash()
+
+            assert key_receive is not None
+            assert nonce_receive is not None
+            assert auth_data is not None
+
+            print("Buffer before decryption:", hexlify(noise_buffer))
+            is_tag_valid = crypto.dec(
+                noise_buffer, tag, key_receive, nonce_receive, auth_data
+            )
```

Comment:
I think `auth_data` should be empty. The [standard](http://www.noiseprotocol.org/noise.html) says:
> Transport messages are then encrypted and decrypted by calling EncryptWithAd() and DecryptWithAd() on the relevant CipherState with zero-length associated data.

Reply:
I'm not sure if this is documented at all in our notion documentation.

Reply:
`handshake hash` was removed from `auth_data` here: https://github.com/trezor/trezor-firmware/pull/4050/commits/0c1d6108bea6a3029ee9bc11a1d4805bcb0390cf